app/controllers/sessions_controller.rb

```diff
@@ -27,10 +27,26 @@ def create
   def twitter
     # raise request.env["omniauth.auth"].to_yaml
     auth = request.env["omniauth.auth"]
-    user = User.where(uid: auth["uid"]).first || User.from_twitter(auth)
-    session[:user_id] = user.id
-    flash[:notice] = "You have been logged in through Twitter."
-    redirect_back_or root_url
+    
+    if signed_in? && @current_user.uid.nil?
```

Comment:
Perfect! A critique would be to say `current_user.update_attribute(:uid, auth.uid).  The only plus side for this is cleaning up the `save` call.  Very well done!